backend/src/main/scala/com/softwaremill/bootzooka/user/UserService.scala

```diff
@@ -158,31 +152,48 @@ class UserService(
   }
 
   private def validatePassword(password: String) =
-    UserRegisterValidator
-      .validatePassword(password)
-      .fold(msg => Fail.IncorrectInput(msg).raiseError[ConnectionIO, Unit], _ => ().pure[ConnectionIO])
+    UserValidator(None, None, Some(password)).asTask()
 }
 
-object UserRegisterValidator {
-  private val ValidationOk = Right(())
+object UserValidator {
   val MinLoginLength = 3
+}
+
+case class UserValidator(loginOpt: Option[String], emailOpt: Option[String], passwordOpt: Option[String]) {
+  private val ValidationOk = Right(())
+
+  private val emailRegex =
+    """^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$""".r
 
-  def validate(login: String, email: String, password: String): Either[String, Unit] =
+  val result: Either[String, Unit] = {
     for {
-      _ <- validateLogin(login.trim)
-      _ <- validateEmail(email.trim)
-      _ <- validatePassword(password.trim)
+      _ <- validateLogin(loginOpt)
+      _ <- validateEmail(emailOpt)
+      _ <- validatePassword(passwordOpt)
     } yield ()
+  }
 
-  def validateLogin(login: String): Either[String, Unit] =
-    if (login.length >= MinLoginLength) ValidationOk else Left("Login is too short!")
+  def asTask(): ConnectionIO[Unit] =
```

Comment:
one more change - that's not really a `Task` we are returning, but a `ConnectionIO`. So this needs renaming, or generalisation into a `def as[F: MonadError]: F[Unit]`